wagtail/admin/panels.py

```diff
@@ -687,20 +734,46 @@ def __repr__(self):
 
     class BoundPanel(Panel.BoundPanel):
         template_name = "wagtailadmin/panels/field_panel.html"
+        read_only_template_name = "wagtailadmin/panels/read_only_field_panel.html"
 
         def __init__(self, **kwargs):
             super().__init__(**kwargs)
 
             if self.form is None:
-                self.bound_field = None
                 return
 
             try:
                 self.bound_field = self.form[self.field_name]
             except KeyError:
                 self.bound_field = None
+
+            if self.read_only:
+                # Ensure heading and help_text are set to something useful
+                self.heading = self.panel.heading or capfirst(
+                    self.panel.db_field.verbose_name
+                )
+                self.help_text = self.panel.help_text or capfirst(
+                    self.panel.db_field.help_text
+                )
+
+                # Remove the field from the form (if present)
```

Comment:
Why would the field be present? Doesn't returning `{}` from `get_form_options` account for that (unless it's being inserted by some other panel, in which case we should probably trust that the other panel knows what it's doing)?

Reply:
When `read_only_on_update` is used, we don't know whether the field is needed until we can tell if the instance is new or not. `get_form_options()` doesn't have that awareness, so we can't do that check there.

Reply:
I was thinking that `FieldPanel.BoundPanel.read_only` might possibly be updated to cater for other cases in future too, such as rendering a read-only value when a `permission` check fails - in which case, you wouldn't be able to figure that out until this point either.

Reply:
Thanks for the clarification! I'll need to spend some time thinking about the best way forward here - the whole premise of the panels refactor in 2.15 was that the Panel object specifies the makeup of the form class it wants to work with, and the BoundPanel then renders the resulting form. If BoundPanel is tinkering with the form fields after the fact, then that design has failed in its aim.

I'm also very reluctant to introduce a new 'hidden' contract between the view and the panels framework about the order things have to happen in. (In fact, if I'm not mistaken - for a successful POST action, we never call get_bound_panel at all, since no template is being rendered in that case, right? Which means it's unavoidable that form validation happens without having first triggered this logic.)

For the `FieldPanel(permission='...')` implementation, I took the stance that if the set of fields is going to vary dynamically, then that custom logic belongs inside the form class rather than in panels, hence the creation of [django-permissionedforms](https://github.com/wagtail/django-permissionedforms). But then that has its own issues - firstly, in this case, we want the form class to have as little involvement as possible with the read-only fields. And secondly, if the form class has to accommodate every conceivable variable that might alter the form (namely, the user in the case of django-permissionedforms, and the create-or-edit flag here), we'll end up with an unmaintainable tower of mixins.

In this case, I think the neatest approach might be to support the create and edit cases using different form classes - i.e. `Panel.get_form_class` and `Panel.get_form_options` will receive the create-or-edit flag and construct a form class tailored to the create or edit case respectively. (Right now we do create a fresh form class on every request, but deliberately limit the information that that process has access to, to avoid the pre-2.15 antipattern where the form class was being assembled on-the-fly and tailored to that specific request, user and object instance. In the current setup, where get_form_options takes no parameters, we're guaranteed that the form class will always be the same for a given model. At some point I'd like to cache those form classes across requests, but haven't got round to it yet... effectively, this change would be relaxing the "form class will always be the same" condition in a very limited and controlled way that leaves open the possibility of caching.)

Reply:
Nicely explained, thank you @gasman. Okay, I will take a look at implementing a `for_create` flag (open to suggestions on naming) for `Panel.get_form_class` and  `Panel.get_form_options`.

Reply:
Resolved in 73a8aa99932588fa568be6ca91d8616d798f6e83